### PR TITLE
feat: make admission webhook CPU over-provisioning configurable

### PR DESCRIPTION
**Pull request by gruizonestic**
## Summary

The admission webhook (Phase 1, `arch-algorithm.md`) provisions every new pod at its **full CPU peak** (`max(SimplePredictionsCPU.MaxValue, CPUStats.P75)`). On capacity-constrained clusters (e.g. a fixed-size node pool) or for workloads that redeploy frequently, this leaves freshly created pods `Pending` until the inflated peak request fits — even though Phase 2 (continuous optimization) trims the running pod back down within minutes.

This PR makes that startup behavior **configurable** while keeping the current behavior as the default.

## What changed

A new setting `recommendationSettings.admissionCPUHeadroomFactor` (env `CRUISEKUBE_RECOMMENDATIONSETTINGS_ADMISSIONCPUHEADROOMFACTOR`) scales how much of the CPU spike headroom (`peak - P75 base`) the admission webhook provisions:

```
request = base(P75) + factor * max(0, peak - base)
```

- `1.0` (**default**) → full peak, **identical to current behavior** (fully backward-compatible)
- `0.0` → steady-state base (P75) only — most aggressive
- e.g. `0.3` → base + 30% of the spike

Memory is intentionally left untouched, since memory under-provisioning risks OOM kills.

The CPU computation is extracted into a pure `admissionCPURequest(base, peak, factor)` helper for testability.

## Files

- `pkg/config/config.go`, `pkg/config/viper.go` — new field + default `1.0`
- `pkg/handlers/controllerwebhookhandler.go` — helper + use it at admission
- `pkg/handlers/controllerwebhookhandler_test.go` — new unit tests + fixture set to `1.0`
- `docs/.../concepts/arch-algorithm.md`, `docs/.../reference/config.md` — documented Phase 1 + the knob
- `config.yaml`, `config.local.yaml` — example value
- `CHANGELOG.md` — Unreleased entry

## Testing

Run locally with the repo's pinned toolchain:

- `go build ./...` — ✅
- `go vet ./...` — ✅
- `go test ./...` — ✅ (new tests + no regression; existing test still yields `300m` with the default factor)
- `gofmt -l` — ✅ clean
- `golangci-lint v2.7.2 run` — ✅ 0 issues

## Notes

- `CONTRIBUTING.md` mentions a `dev` branch, but it does not currently exist on the repo, so this PR targets `main`. Happy to retarget if needed.

---

> [!NOTE]
> **Medium Risk**
> Changes admission webhook CPU requests for all newly created pods when factor is below 1.0, which can increase Pending scheduling or reliance on Phase 2 in-place resize; default 1.0 preserves prior behavior.
> 
> **Overview**
> Adds **`recommendationSettings.admissionCPUHeadroomFactor`** (default **`1.0`**) so Phase 1 admission CPU requests use **`P75 + factor × (peak − P75)`** instead of always provisioning the full CPU peak. Memory admission logic is unchanged.
> 
> The mutating webhook now calls **`admissionCPURequest`** with stats P75 and predicted peak; config, Viper default, sample YAML, changelog, and algorithm/config docs describe tuning and the **`CRUISEKUBE_RECOMMENDATIONSETTINGS_ADMISSIONCPUHEADROOMFACTOR`** env var. Unit tests cover factor values and edge cases (negative factor, peak below base).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 25dc19b0094d2778cfd7d5580e7ffd734a178c12. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added configurable CPU headroom for admission recommendations.
  * CPU requests now combine steady-state usage with an adjustable portion of predicted spike capacity.
  * The default setting preserves full peak headroom.

* **Documentation**
  * Documented the new setting, configuration options, tuning examples, and scheduling tradeoffs.

* **Tests**
  * Added coverage for headroom settings, fallback behavior, and negative-value handling.

### CODE REVIEW

**Comment by gruizonestic** [comment]
Thanks @coderabbitai — both points addressed in the latest push:

1. **Peak-only wording (arch-algorithm.md, Purpose + "Preventing Repeated Evictions")**: reworded to make the peak guarantee conditional on `admissionCPUHeadroomFactor == 1.0`, and noted that lower factors admit pods at `base + factor * (peak - base)` and may need an in-place upward resize once Phase 2 runs.
2. **MD046 (indented code block in the tip admonition)**: replaced the `!!! tip` admonition with a regular `###` subsection and a flush-left bullet list. Verified locally with markdownlint-cli2 v0.18.1 — no MD046 on the file.

Also dropped some redundant inline comments from the Go/YAML changes per maintainer preference.

**Comment by shubhamrai1993** [comment]
Hi @gruizonestic , thanks for contributing. Will take a look. 

Can you add more color to the use case. We have been running this with auto-provisioned clusters with karpenter. Without auto-provisioning, the cluster will not delete any node or reclaim any space after requests being decreased. That will anyways not save any cost. So does the tool even make sense?

**Comment by gruizonestic** [comment]
Hi @shubhamrai1993, thanks for taking a look!

Fair point — you're right that on a non-auto-provisioned cluster, lowering requests won't scale nodes down, so it doesn't cut the cloud bill directly. The value on a **fixed cluster** is different:

1. Bin-packing density. Accurate requests let me fit more workloads onto the same nodes. On my cluster (3 nodes, ~80% utilization) I genuinely couldn't fit everything before CruiseKube — the Phase 2 in-place right-sizing is what makes it fit. So it's "do more with the nodes I already pay for" rather than "pay for fewer nodes."

2. Operability of a packed cluster — this is what the PR targets. We get frequent Kubernetes upgrades that drain nodes and reschedule everything. The admission webhook sets each new pod's request to `max(SimplePredictionsCPU.MaxValue, CPUStats.P75)`, and since the peak `(MaxValue)` is usually well above P75, every workload is admitted asking for far more CPU than it needs at startup. On an already-packed cluster nothing can reschedule, so the cluster effectively can't redeploy. With auto-provisioning this is invisible (Karpenter just adds a node); without it, it's a hard wall.

The PR keeps full-peak provisioning as the default `(admissionCPUHeadroomFactor: 1.0)`, so it's a no-op for your Karpenter setups. Lowering the factor only helps capacity-constrained clusters: pods are admitted at `base + factor*(peak - base)` and Phase 2 still right-sizes them once running — so no optimization is lost, you just avoid over-provisioning at the worst possible moment (a mass reschedule). 

Happy to add this "constrained cluster" use case to the docs / PR description if useful. And if you did see the change affect auto-provisioning in your testing, point me at the scenario and I'll dig in.

**Comment by shubhamrai1993** [comment]
Hi looking at this PR right now. It seems to have conflicts. Can you resolve them please?

**Comment by gruizonestic** [comment]
Hi @shubhamrai1993! Conflicts solved. Let me know if you need something else.

File: pkg/handlers/controllerwebhookhandler.go [hunk 1]
```diff
@@ -463,6 +463,17 @@ func (deps HandlerDependencies) adjustResources(ctx context.Context, pod *corev1
 	return patches
 }
 
+func admissionCPURequest(base, peak, factor float64) float64 {
+	headroom := peak - base
+	if headroom < 0 {
+		headroom = 0
+	}
+	if factor < 0 {
+		factor = 0
```
Comment (shubhamrai1993) [inline pkg/handlers/controllerwebhookhandler.go:472]: factor should never be less than 0. should get discarded at the configuration validation layer

File: pkg/config/config.go [hunk 2]
```diff
@@ -105,6 +105,8 @@ type RecommendationSettings struct {
 	MaxConcurrentQueries      int  `yaml:"maxConcurrentQueries" mapstructure:"maxConcurrentQueries"`
 	OOMCooldownMinutes        int  `yaml:"oomCooldownMinutes" mapstructure:"oomCooldownMinutes"`
 	OptimizeGuaranteedPods    bool `yaml:"optimizeGuaranteedPods" mapstructure:"optimizeGuaranteedPods"`
+
+	AdmissionCPUHeadroomFactor float64 `yaml:"admissionCPUHeadroomFactor" mapstructure:"admissionCPUHeadroomFactor"`
```
Comment (shubhamrai1993) [inline pkg/config/config.go:109]: why add an extra line here?

File: pkg/handlers/controllerwebhookhandler.go (hunk 1 above)
Comment (shubhamrai1993) [reply pkg/handlers/controllerwebhookhandler.go:472]: should we also clamp factor > 1.0?
